File: android/src/main/java/com/tencent/qcloud/tuiplayer/flutter/ftuiplayer_kit/player/FTUIShortController.java
package com.tencent.qcloud.tuiplayer.flutter.ftuiplayer_kit.player;

import android.content.Context;

import androidx.annotation.NonNull;

import com.tencent.qcloud.tuiplayer.core.api.TUIPlayerBridge;
import com.tencent.qcloud.tuiplayer.core.api.TUIPlayerManager;
import com.tencent.qcloud.tuiplayer.core.api.TUIPlayerVodStrategy;
import com.tencent.qcloud.tuiplayer.core.api.model.TUIPlaySource;
import com.tencent.qcloud.tuiplayer.core.api.model.TUIVideoSource;
import com.tencent.qcloud.tuiplayer.core.api.tools.TUIDataUtils;
import com.tencent.qcloud.tuiplayer.core.api.ui.player.ITUIVodPlayer;
import com.tencent.qcloud.tuiplayer.core.api.ui.view.TUIBaseVideoView;
import com.tencent.qcloud.tuiplayer.core.preload.TUIVideoDataHolder;
import com.tencent.qcloud.tuiplayer.core.tools.TUIPlayerLog;
import com.tencent.qcloud.tuiplayer.flutter.ftuiplayer_kit.messages.FTUIMessages;
import com.tencent.qcloud.tuiplayer.flutter.ftuiplayer_kit.tools.FTUITransformer;
import com.tencent.qcloud.tuiplayer.flutter.ftuiplayer_kit.view.FTUIItemViewFactory;
import com.tencent.qcloud.tuiplayer.flutter.ftuiplayer_kit.view.FTUIShortVideoItemView;

import java.util.List;

import io.flutter.plugin.common.BinaryMessenger;

public class FTUIShortController implements TUIPlayerBridge, FTUIMessages.FTUIPlayerShortAPI {

    private static final String TAG = "FTUIShortController";

    private final TUIPlayerManager mManager;
    private final FTUIItemViewFactory mViewFactory;
    private final int mId;
    private final FTUIShortEngineObserver mEngineObserver;
    private int mCurrentIndex;
    private final TUIVideoDataHolder mDataHolder;
    private final BinaryMessenger mMessenger;
    private boolean mIsOneLoop = true;

    public FTUIShortController(Context context, FTUIItemViewFactory viewFactory, int id, BinaryMessenger messenger,
                               FTUIShortEngineObserver observer) {
        TUIPlayerLog.i(TAG, "start create shortController,controllerId " + id);
        mMessenger = messenger;
        mManager = new TUIPlayerManager(context, this);
        mDataHolder = mManager.getDataHolder();
        mViewFactory = viewFactory;
        mId = id;
        mEngineObserver = observer;
        FTUIMessages.FTUIPlayerShortAPI.setUp(messenger, String.valueOf(mId), this);
    }

    public void bindVideoView(int viewId, boolean isPreBind, int curIndex) {
        FTUIShortVideoItemView itemView = mViewFactory.findViewById(viewId);
        final int count = mDataHolder.size();
        if (curIndex >= count || curIndex < 0) {
            TUIPlayerLog.e(TAG, "bindVideoView failed, index outOfRange,index:"
                    + curIndex + ",isPreBind:" + isPreBind);
            return;
        }
        if (null != itemView) {
            TUIBaseVideoView itemVideoView = (TUIBaseVideoView) itemView.getView();
            itemVideoView.bindVideoModel(mDataHolder.getSource(curIndex));
            if (isPreBind) {
                TUIPlayerLog.v(TAG, "start preRender index:" + curIndex);
                if (Math.abs(curIndex - mCurrentIndex) > 2) {
                    TUIPlayerLog.w(TAG, "jump preRender index:" + curIndex
                            + ", the difference between the index and the current coordinates is too large");
                } else {
                    mManager.preRenderOnView(itemVideoView);
                }
            } else {
                mCurrentIndex = curIndex;
                TUIPlayerLog.v(TAG, "mCurrentIndex start update to:" + mCurrentIndex);
                mManager.bindVideoView(itemVideoView);
                TUIPlayerLog.v(TAG, "mCurrentIndex updated to:" + mCurrentIndex);
            }
            handlePlayerLoopMode(itemVideoView);
        } else {
            TUIPlayerLog.e(TAG, "bindVideoView met a null view,viewId:" + viewId + ",isPreBind:" + isPreBind);
        }
    }

    @NonNull
    @Override
    public Long setModels(@NonNull FTUIMessages.FTUIListVodSourceMsg msg) {
        List<TUIPlaySource> sources = TUIDataUtils.copyModels(FTUITransformer.transToListVodSourceFromMsg(msg));
        return (long) mManager.setModels(sources);
    }

    @NonNull
    @Override
    public Long appendModels(@NonNull FTUIMessages.FTUIListVodSourceMsg msg) {
        List<TUIPlaySource> copyModels = TUIDataUtils.copyModels(FTUITransformer.transToListVodSourceFromMsg(msg));
        return (long) mManager.appendModels(copyModels);
    }

    @NonNull
    @Override
    public Long startCurrent() {
        return (long) mManager.startCurrent();
    }

    @Override
    public void setVodStrategy(@NonNull FTUIMessages.FTUIPlayerVodStrategyMsg msg) {
        TUIPlayerVodStrategy strategy = FTUITransformer.transToVodStrategyOpenPreFromMsg(msg);
        mManager.updateVodStrategy(strategy);
    }

    @NonNull
    @Override
    public FTUIMessages.FTUIVodSourceMsg getCurrentModel() {
        TUIPlaySource source = mManager.getCurrentModel();
        if (source instanceof TUIVideoSource) {
            return FTUITransformer.transToMsgFromVodSource((TUIVideoSource) source);
        }
        return new FTUIMessages.FTUIVodSourceMsg();
    }

    @Override
    public void bindVideoView(@NonNull Long pageViewId, @NonNull Long index) {
        bindVideoView(pageViewId.intValue(), false, index.intValue());
    }

    @Override
    public void preBindVideo(@NonNull Long pageViewId, @NonNull Long index) {
        bindVideoView(pageViewId.intValue(), true, index.intValue());
    }

//    @Override
//    public void removeModels(@NonNull Long startIndex, @NonNull Long itemCount) {
//        mManager.removeModels(startIndex.intValue(), itemCount.intValue());
//    }
//
//    @Override
//    public void insertModels(@NonNull FTUIMessages.FTUIListVodSourceMsg msg, @NonNull Long startIndex) {
//        List<TUIPlaySource> sources = TUIDataUtils.copyModels(FTUITransformer.transToListVodSourceFromMsg(msg));
//        mManager.insertModels(sources, startIndex.intValue());
//    }
//
//    @Override
//    public void replaceModel(@NonNull FTUIMessages.FTUIVodSourceMsg msg, @NonNull Long index) {
//        TUIPlaySource source = FTUITransformer.transToVodSourceFromMsg(msg);
//        mManager.replaceModel(source, index.intValue());
//    }

    private void handlePlayerLoopMode(TUIBaseVideoView itemView) {
        if (null != itemView && null != itemView.getController() && null != itemView.getController().getPlayer()) {
            itemView.getController().setLoop(mIsOneLoop);
        }
    }

    public void release() {
        TUIPlayerLog.i(TAG, "start release shortController,controllerId " + mId);
        FTUIMessages.FTUIPlayerShortAPI.setUp(mMessenger, String.valueOf(mId), null);
        mManager.releasePlayers();
        mEngineObserver.onRelease(mId);
    }

    @Override
    public void setVideoLoop(@NonNull Boolean isLoop) {
        this.mIsOneLoop = isLoop;
    }


    @Override
    public int getCurrentScrollState() {
        return 0;
    }

    @Override
    public int getCurrentPlayingIndex() {
        return mCurrentIndex;
    }

    @Override
    public void onCurrentPlayEnd() {

    }

    @Override
    public void postHandlePlayCurrent(int i) {

    }

    @Override
    public void postOnMain(Runnable runnable) {

    }

    @Override
    public void onVodPlayerReady(ITUIVodPlayer ituiVodPlayer, TUIVideoSource tuiVideoSource) {

    }

    @Override
    public void changeCurPos(int i) {

    }
}

File: ios/Classes/player/FTUIShortController.m
//
//  FTUIShortController.m
//  ftuiplayer_kit
//
//  Created by kongdywang on 21.8.24.
//

#import "FTUIShortController.h"
#import "FtxMessages.h"
#import <TUIPlayerCore/TUIPlayerCore-umbrella.h>
#import "FTUITransformer.h"
#import "FTUIConstant.h"

@interface FTUIShortController() <FTUIPlayerShortAPI>

@property (nonatomic, strong) TUIPlayerVodStrategyManager *vodStrategyManager;///策略管理
@property (nonatomic, strong) TUIPlayerVodManager* vodManager;
@property (nonatomic, strong) TUIPlayerVodPreLoadManager* preloadManager;
@property (nonatomic, strong) TUIPlayerRecordManager* recordManager;
@property (nonatomic, strong) NSMutableArray* dataArray;
@property (nonatomic, assign) long currentIndex;
@property (nonatomic, assign) long preIndex;
@property (nonatomic, strong) FTUIVodController *curVodController;

@property (nonatomic, strong) FTUIItemViewFactory* viewFactory;
@property (nonatomic, strong) id<FTUIShortEngineObserver> engineObserver;
@property (nonatomic, assign) NSUInteger controllerId;

@property (nonatomic, assign) long cachePageIndex;
@property (nonatomic, assign) long cacheIndex;

@end

@implementation FTUIShortController

- (nonnull instancetype)initWithViewFactory:(nonnull FTUIItemViewFactory *)viewFactory controllerId:(NSUInteger)controllerId messenger:(nonnull id<FlutterBinaryMessenger>)messenger observer:(nonnull id<FTUIShortEngineObserver>)observer {
    self = [super init];
    TUILOGI(@"start create shortController,controllerId %lu", controllerId)
    if (self) {
        self.vodStrategyManager = [[TUIPlayerVodStrategyManager alloc] init];
        self.dataArray = @[].mutableCopy;
        self.currentIndex = -1;
        self.preIndex = -1;
        self.engineObserver = observer;
        self.viewFactory = viewFactory;
        self.controllerId = controllerId;
        self.curVodController = nil;
        self.cachePageIndex = -1;
        self.cacheIndex = -1;
        SetUpFTUIPlayerShortAPIWithSuffix(messenger, self, [NSString stringWithFormat:@"%lu", controllerId]);
    }
    return self;
}

///清除点播播放器缓存
- (void)removeVodPlayerCache:(long)curIndex preIndex:(long)preIndex {
    @synchronized (self.dataArray) {
        TUILOGI(@"[removeVodPlayerCache] currentIndex:%ld", curIndex)
        bool isDonwFlip = (curIndex - preIndex) >= 0;
        ///  flutter 端的 viewPager，由于会先 itemBuilder，所以会先触发预播放，然后再触发当前页面的 bind，
        ///  由于 IOS 取播放器的时候，不会触发 LRU 的顺序更新，所以三播放器的情况下，会导致next 的播放器经常无法获取到
        ///如果开启了预播放上一个，那么上一个的播放器保持，只清除上上个播放器缓存
        ///如果没有开启，那么上一个播放器缓存及时清除
        long prePreIndex =  curIndex - 2;
        long preIndex = curIndex - 1;
        long nextIndex = curIndex + 1;
        long nextNextIndex = curIndex + 2;
        [self removeVodPlayerCacheByIndex:curIndex - 3];
        [self removeVodPlayerCacheByIndex:prePreIndex];
        [self removeVodPlayerCacheByIndex:nextNextIndex];
        if (isDonwFlip) {
            [self removeVodPlayerCacheByIndex:preIndex];
        } else {
            [self removeVodPlayerCacheByIndex:nextIndex];
        }
    }
}

-(void)removeVodPlayerCacheByIndex:(long)removeIndex {
    NSUInteger dataCount = self.dataArray.count;
    TUIPlayerDataModel *removeModel = nil;
    if (dataCount > removeIndex && removeIndex >= 0) {
        removeModel = self.dataArray[removeIndex];
    }
    if (removeModel && removeModel.modelType == TUI_MODEL_TYPE_VOD) {
        [self.vodManager removePlayerCache:removeModel.asVodModel];
    }
}

- (void)bindVideoView:(NSInteger)pageViewId index:(NSInteger)index isPreBind:(BOOL)isPreBind {
    NSUInteger dataCount = self.dataArray.count;
    if (index < 0 || index >= dataCount) {
        TUILOGE(@"bindVideoView failed, index outOfRange,index:%lu, isPreBind:%d", index, isPreBind)
        return;
    }
    FTUIShortVideoItemView *itemView = [self.viewFactory findViewById:pageViewId];
    if (nil != itemView) {
        TUIPlayerDataModel *dataModel = self.dataArray[index];
        if ([dataModel isKindOfClass:[TUIPlayerVideoModel class]]) {
            TUIPlayerVideoModel* vodModel = [dataModel asVodModel];
            [self.vodManager.vodPreLoadManager cancelPreLoadOperationWith:vodModel];
            __block TUIShortVideoItemView* videoItemView = (TUIShortVideoItemView*)[itemView view];
            [videoItemView setItemViewModel:vodModel];
            if (isPreBind) {
                TUILOGI(@"start preRender index:%lu", index)
                if (self.currentIndex == index && [self.vodManager currentVodPlayer]) {
                    TUITXVodPlayer *player = [self.vodManager currentVodPlayer];
                    if (player.status < TUITXVodPlayerStatusPrepared && player.status > TUITXVodPlayerStatusEnded) {
                        TUILOGW(@"prePlay a idle player,index:%lu", index)
                        [self.vodManager prePlayWithModel:vodModel type:1];

                        [self.vodManager setVideoWidget:videoItemView.videoBaseView.videoContainerView model:vodModel
                                     firstFrameCallBack:^(BOOL isFirstFrame) {
                            [videoItemView hiddenCoverImage:YES];
                        }];
                    } else {
                        TUILOGW(@"video is playing, jump prePlay opt,index:%lu", index)
                    }
                }
                else if (ABS(self.currentIndex - index) > 2) {
                    TUILOGW(@"jump preRender index:%lu, the difference between the index and the current coordinates is too large", self.currentIndex);
                }
                else {
                    TUILOGW(@"start prePlay a player, index:%lu", index)
                    [self.vodManager prePlayWithModel:vodModel type:1];

                    [self.vodManager setVideoWidget:videoItemView.videoBaseView.videoContainerView model:vodModel
                                 firstFrameCallBack:^(BOOL isFirstFrame) {
                        [videoItemView hiddenCoverImage:YES];
                    }];
                }
            } else {
                [self.vodManager resetAllPlayer];
                [self.vodManager muteAllPlayer];
                TUILOGI(@"mCurrentIndex start update to:%lu", index)
                if (self.curVodController != nil) {
                    [self.curVodController onUnBindController];
                    [self.vodManager removeDelegate:self.curVodController];
                    self.curVodController = nil;
                }
                [self.vodManager addDelegate:[itemView getVodController]];
                [self.vodManager.vodPreLoadManager setCurrentPlayingModel:vodModel];
                [self.vodManager playWithModel:vodModel];
                [self removeVodPlayerCache:index preIndex:self.currentIndex];
                self.curVodController = [itemView getVodController];
                [[itemView getVodController] onBindController:self.vodManager.currentVodPlayer];
                self.preIndex = self.currentIndex;
                self.currentIndex = index;
                TUILOGI(@"mCurrentIndex updated to:%lu", self.currentIndex)
                [self.vodManager setVideoWidget:videoItemView.videoBaseView.videoContainerView model:vodModel
                             firstFrameCallBack:^(BOOL isFirstFrame) {
                    [videoItemView hiddenCoverImage:YES];
                }];
                [videoItemView getPlayer:self.vodManager.currentVodPlayer];
                if (index == 0) {
                    self.cachePageIndex = 1;
                    self.cacheIndex = 1;
                }
                if (self.cachePageIndex >= 0 && self.cacheIndex >= 0) {
                    [self bindVideoView:self.cachePageIndex index:self.cacheIndex isPreBind:YES];
                    self.cachePageIndex = -1;
                    self.cacheIndex = -1;
                }
            }
        } else {
            TUILOGE(@"unImpl modelType:%@", dataModel);
        }
    } else {
        TUILOGE(@"bindVideoView met a null view,viewId:%lu,isPreBind:%i", pageViewId, isPreBind);
    }
}

- (void)bindVideoViewPageViewId:(NSInteger)pageViewId index:(NSInteger)index error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [self bindVideoView:pageViewId index:index isPreBind:NO];
}

- (void)preBindVideoPageViewId:(NSInteger)pageViewId index:(NSInteger)index error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    self.cachePageIndex = pageViewId;
    self.cacheIndex = index;
//    [self bindVideoView:pageViewId index:index isPreBind:YES];
}

- (nullable NSNumber *)setModelsMsg:(nonnull FTUIListVodSourceMsg *)msg error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    @synchronized (self.dataArray) {
        self.currentIndex = 0;
        self.preIndex = 0;
        [self.vodManager removeAllPlayerCache];
        [self.vodManager stopAllPlayer];
        self.vodManager.currentVodPlayer = nil;
        NSArray *internalModels = [FTUITransformer transToListVodSourceFromMsg:msg];
        [self.dataArray removeAllObjects];
        [self.dataArray addObjectsFromArray:internalModels];
        [self.vodManager.vodPreLoadManager setPlayerModels:internalModels];
        [self.vodManager.vodDataManager setShortVideoModels:internalModels];
        return @(0);
    }
}

- (nullable NSNumber *)appendModelsMsg:(nonnull FTUIListVodSourceMsg *)msg error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    @synchronized (self.dataArray) {
        NSArray *internalModels = [FTUITransformer transToListVodSourceFromMsg:msg];
        [self.dataArray addObjectsFromArray:internalModels];
        [self.vodManager.vodPreLoadManager appendPlayerModels:internalModels];
        [self.vodManager.vodDataManager appendShortVideoModels:internalModels];
        TUIPlayerVideoModel *nextModel = [self getVodModel:self.currentIndex + 1];
        if (nil != nextModel) {
            [self.preloadManager cancelPreLoadOperationWith:nextModel];
            [self.vodManager prePlayWithModel:nextModel type:1];
        }
        return @(0);
    }
}

- (TUIPlayerVideoModel*)getVodModel:(NSUInteger)index {
    TUIPlayerVideoModel* model = nil;
    if (index >= 0 && index < self.dataArray.count) {
        TUIPlayerDataModel* dataModel = self.dataArray[index];
        if (nil != dataModel && [dataModel isKindOfClass:[TUIPlayerVideoModel class]]) {
            model = dataModel.asVodModel;
        }
    }
    return model;
}

- (nullable FTUIVodSourceMsg *)getCurrentModelWithError:(FlutterError * _Nullable __autoreleasing * _Nonnull)error 
{
    TUIPlayerVideoModel* vodModel = [self getVodModel:self.currentIndex];
    if (nil != vodModel) {
        return [FTUITransformer transToMsgFromVodSource:vodModel];
    }
    return [[FTUIVodSourceMsg alloc] init];
}

- (void)releaseWithError:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    TUILOGI(@"start release shortController,controllerId:%lu", self.controllerId)
    [self.vodManager.currentVodPlayer removeVideo];
    self.vodManager.currentVodPlayer = nil;
    [self.vodManager stopAllPlayer];
    [self.engineObserver onRelease:self.controllerId];
}

- (void)setVodStrategyMsg:(nonnull FTUIPlayerVodStrategyMsg *)msg error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    dispatch_async(dispatch_get_main_queue(), ^{
        TUIPlayerVodStrategyModel* vodStrategyModel = [FTUITransformer transToVodStrategyOpenPreFromMsg:msg];
        [self.vodStrategyManager setVideoStrategy:vodStrategyModel];
        [self.vodManager setVodStrategyManager:self.vodStrategyManager];
        self.vodManager.vodPreLoadManager.strategyManager = self.vodStrategyManager;
    });
}

- (nullable NSNumber *)startCurrentWithError:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    BOOL result = tuicpa();
    if (result == NO) {
        TUILOGE(@"[setModels] License checked failed! Player Premium license required!")
        return @(TUI_ERROR_INVALID_LICENSE);
    }
    if (nil != self.vodManager.currentVodPlayer) {
        [self.vodManager.currentVodPlayer resumePlay];
    }
    return @(0);
}

- (TUIPlayerRecordManager*)recordManager {
    if (self->_recordManager == nil) {
        self->_recordManager = [[TUIPlayerRecordManager alloc] init];
        self->_recordManager.strategyManager = self.vodStrategyManager;
    }
    return self->_recordManager;
}

- (TUIPlayerVodManager*)vodManager {
    if (self->_vodManager == nil) {
        self->_vodManager = [[TUIPlayerVodManager alloc] init];
        [self->_vodManager setVodStrategyManager:self.vodStrategyManager];
    }
    return self->_vodManager;
}

- (TUIPlayerVodPreLoadManager*)preloadManager {
    if (self->_preloadManager == nil) {
        self->_preloadManager = [[TUIPlayerVodPreLoadManager alloc] init];
        self->_preloadManager.strategyManager = self.vodStrategyManager;
    }
    return self->_preloadManager;
}

#pragma mark - cpa
static BOOL tuicpa(void){
    Class tpaClass = NSClassFromString(@"TUIPlayerAuth");
    SEL cpaSelector = NSSelectorFromString(@"cpa");
    IMP cpaImp = [tpaClass methodForSelector:cpaSelector];
    BOOL (*cpaFunc)(id, SEL) = (BOOL (*)(id, SEL))cpaImp;
    BOOL result = cpaFunc(tpaClass, cpaSelector);
    return result;
}
@end
